astropy/coordinates/polarization.py

```diff
@@ -0,0 +1,258 @@
+    def unit(self):
+        return None
+
+    @property
+    def dtype(self):
+        return self._parent._data.dtype
+
+    @staticmethod
+    def default_format(val):
+        return f"{val.symbol}"
+
+    def new_like(self, cols, length, metadata_conflicts="warn", name=None):
+        """
+        Return a new StokesCoord instance which is consistent with the
+        input ``cols`` and has ``length`` rows.
+
+        This is intended for creating an empty column object whose elements can
+        be set in-place for table operations like join or vstack.
+
+        Parameters
+        ----------
+        cols : list
+            List of input columns
+        length : int
+            Length of the output column object
+        metadata_conflicts : str ('warn'|'error'|'silent')
+            How to handle metadata conflicts
+        name : str
+            Output column name
+
+        Returns
+        -------
+        col : `~astropy.coordinates.StokesCoord` (or subclass)
+            Empty instance of this class consistent with ``cols``
+
+        """
+        # Get merged info attributes like shape, dtype, format, description, etc.
+        attrs = self.merge_cols_attributes(
+            cols, metadata_conflicts, name, ("meta", "format", "description")
+        )
+
+        # Make an empty StokesCoord.
+        shape = (length,) + attrs.pop("shape")
+        data = np.zeros(shape=shape, dtype=attrs.pop("dtype"))
+        # Get arguments needed to reconstruct class
+        out = self._construct_from_dict(dict(value=data))
+
+        # Set remaining info attributes
+        for attr, value in attrs.items():
+            setattr(out.info, attr, value)
+
+        return out
+
+    def get_sortable_arrays(self):
+        """
+        Return a list of arrays which can be lexically sorted to represent
+        the order of the parent column.
+
+        For StokesCoord this is just the underlying values.
+
+        Returns
+        -------
+        arrays : list of ndarray
+        """
+        return [self._parent._data]
+
+
+class StokesCoord(ShapedLikeNDArray):
+    """
+    A representation of stokes coordinates with helpers for converting to profile names.
+
+    Parameters
+    ----------
+    stokes_values : array-like
+        The numeric values representing stokes coordinates.
+    """
+
+    info = StokesCoordInfo()
+
+    def __init__(self, stokes, copy=False):
+        if isinstance(stokes, type(self)):
+            self._data = stokes._data.copy() if copy else stokes._data
+            self.info = stokes.info
+        else:
+            data = np.asanyarray(stokes)
+
+            if data.dtype.kind == "O":
+                msg = "StokesCoord objects cannot be initialised with an object array."
+                raise ValueError(msg)
+
+            if data.dtype.kind == "U":
+                self._data = self._from_symbols(data)
+            else:
+                self._data = data.copy() if copy and data is stokes else data
```

Comment:
Data is always assigned, so maybe the following is more explicit
```suggestion
        if isinstance(stokes, type(self)):
            data = stokes._data.copy() if copy else stokes._data
            self.info = stokes.info
        else:
            data = np.asanyarray(stokes)

            if data.dtype.kind == "O":
                msg = "StokesCoord objects cannot be initialised with an object array."
                raise ValueError(msg)

            if data.dtype.kind == "U":
                data = self._from_symbols(data)
            else:
                data = data.copy() if copy and data is stokes else data
        self._data = data
```

Reply:
If the dataclass suggestion is used this would be

```
object.__setattr__(self "data", data)
```